core/src/main/java/org/mule/runtime/core/internal/event/DefaultEventContext.java

```diff
@@ -331,7 +348,7 @@ public String getId() {
 
     @Override
     public String getCorrelationId() {
-      return parent.getCorrelationId();
+      return correlationId.orElse(parent.getCorrelationId());
```

Comment:
why don't you do this directly in the constructor and avoid having the `Optional`?